core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java

```diff
@@ -227,6 +230,19 @@ private void configureAsHttps(Channel ch, SocketAddress remoteAddr) {
             sslHandler = sslCtx.newHandler(ch.alloc());
         }
 
+        final boolean fastOpen = Boolean.TRUE.equals(ch.config().getOption(ChannelOption.TCP_FASTOPEN_CONNECT));
+        if (!fastOpen) {
+            p.addLast(new ChannelInboundHandlerAdapter() {
+                @Override
+                public void channelActive(ChannelHandlerContext ctx) throws Exception {
+                    final ClientConnectionTimingsBuilder timingsBuilder =
+                            ctx.channel().attr(TIMINGS_BUILDER_KEY).get();
+                    timingsBuilder.tlsHandshakeStart();
+                    super.channelActive(ctx);
+                    p.remove(this);
+                }
```

Comment:
I think it's probably better to unify TLS related logic to a single handler.

What do you think of moving this logic to the following location? This way, we don't have to increase complexity using another handler.

https://github.com/line/armeria/blob/2164adbb053a4df50e5f1108f314637c208ac8be/core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java#L250

If the order of `tlsHandshakeStart` and `tlsHandshakeEnd` is an issue, I think it is better to handle this inside `ClientConnectionTimingsBuilder`

Reply:
I suggested starting the measurement in either the 1st or 4th handler, depending on how long it takes for the active call to reach the 4th handler:
- If the duration is relatively short, place the measurement in the 4th handler to unify the logic.
- If the duration is long, place it in the 1st handler.

@Leewongi0731 informed me that it takes more than 10 milliseconds, so I agree that we should put the measurement in the 1st handler. Additionally, the handlers are removed after use, and they are all added within the same method, which is pretty unified. 😉 

@Leewongi0731 Could you provide the numbers to the devs as well?

Reply:
I ran the code below and the `httpsServerConnectionWithTlsSelfSigned()` unit test to measure the time.

```java
p.addLast(new ChannelInboundHandlerAdapter() {
  public void channelActive(ChannelHandlerContext ctx) throws Exception {
    System.out.println("[TEST_LOG] first handler channelActive() called : " + SystemInfo.currentTimeMicros());
  }
}); // 1st handler
p.addLast(configureSslHandler(sslHandler));          // 2nd handler
p.addLast(TrafficLoggingHandler.CLIENT);             // 3rd handler
p.addLast(new ChannelInboundHandlerAdapter() {
  public void channelActive(ChannelHandlerContext ctx) throws Exception {
    System.out.println("[TEST_LOG] fourth handler channelActive() called : " + SystemInfo.currentTimeMicros());
  }
}); // 4th handler
```

```
[TEST_LOG] first handler channelActive() called : 1714562282289000
[TEST_LOG] fourth handler channelActive() called : 1714562282298000
=> 1714562282298000 - 1714562282289000 = 9000 ns = 0.009 ms

[TEST_LOG] first handler channelActive() called : 1714562441837000
[TEST_LOG] fourth handler channelActive() called : 1714562441844000
=> 1714562441844000 - 1714562441837000 = 7000 ns = 0.007 ms

[TEST_LOG] first handler channelActive() called : 1714562510353000
[TEST_LOG] fourth handler channelActive() called : 1714562510360000
=> 1714562510360000 - 1714562510353000 = 7000 ns = 0.007 ms
```

There was definitely a 10ms~11ms difference when I tested it before, but now there is just difference of 0.01ms.. 😂
If this difference is this much, I agreed to the integrate it into single handler! If there is a way to do a more advanced test, please give me opinion.

Reply:
Could you check it again using `System.nanoTime();`, please?

Reply:
If it takes approximately 10 ms, we should do this before `SslHandler.channelActive()` is invoked. 😉 
@trustin suggested that we can just extend `SslHandler` and collect the metric there:
```java
final class ClientSslHandler extends SslHandler {
    private final ClientConnectionTimingsBuilder timingsBuilder;

    ClientSslHandler(SSLEngine engine, ClientConnectionTimingsBuilder timingsBuilder) {
        super(engine);
        this.timingsBuilder = timingsBuilder;
    }

    @Override
    public void channelActive(ChannelHandlerContext ctx) throws Exception {
        timingsBuilder.tlsHandshakeStart();
        super.channelActive(ctx);
        handshakeFuture().addListener(future -> {
            timingsBuilder.tlsHandshakeEnd();
        });
    }
}

...
p.addLast(configureSslHandler(new ClientSslHandler(sslCtx.newEngine(ch.alloc()), timingsBuilder)));
```